src/hotspot/share/runtime/threadSMR.cpp

```diff
@@ -193,29 +193,26 @@ class ThreadScanHashtable : public CHeapObj<mtThread> {
   }
 
   int _table_size;
-  // ResourceHashtable SIZE is specified at compile time so our
-  // dynamic _table_size is unused for now; 1031 is the first prime
-  // after 1024.
-  typedef ResourceHashtable<void *, int, &ThreadScanHashtable::ptr_hash,
-                            &ThreadScanHashtable::ptr_equals, 1031,
-                            ResourceObj::C_HEAP, mtThread> PtrTable;
-  PtrTable * _ptrs;
+  KVHashtable<void *, bool, mtThread, &ThreadScanHashtable::ptr_hash,
+                            &ThreadScanHashtable::ptr_equals> _ptrs;
 
  public:
-  // ResourceHashtable is passed to various functions and populated in
-  // different places so we allocate it using C_HEAP to make it immune
-  // from any ResourceMarks that happen to be in the code paths.
-  ThreadScanHashtable(int table_size) : _table_size(table_size), _ptrs(new (ResourceObj::C_HEAP, mtThread) PtrTable()) {}
+  ThreadScanHashtable(int table_size) : _table_size(table_size), _ptrs(table_size) {
+      log_trace(thread, smr)("tid=" UINTX_FORMAT ": allocate ThreadScanHashtable(%d) at " INTPTR_FORMAT, os::current_thread_id(), _table_size, p2i(this));
+  }
 
-  ~ThreadScanHashtable() { delete _ptrs; }
+  ~ThreadScanHashtable() {
+      log_trace(thread, smr)("tid=" UINTX_FORMAT ": deallocate ThreadScanHashtable(%d) at " INTPTR_FORMAT, os::current_thread_id(), _table_size, p2i(this));
```

Comment:
For a CHeap hashtable, I believe this should do this (or have KVHash have an API to delete all the entries).
Dictionary::~Dictionary() {
  DictionaryEntry* probe = NULL;
  for (int index = 0; index < table_size(); index++) {
    for (DictionaryEntry** p = bucket_addr(index); *p != NULL; ) {
      probe = *p;
      *p = probe->next();
      free_entry(probe);
    }
  }
  assert(number_of_entries() == 0, "should have removed all entries");
}

Reply:
This should be added to KVHashtable because it would be nice to migrate the other hashtables to this nicer API.

Reply:
I don't see how the entries in the table are freed (in any of the KVHashtables).  Maybe I'm missing something.

Reply:
I'll take a look.

Reply:
https://bugs.openjdk.java.net/browse/JDK-8267752